x/wasm/internal/keeper/ibc.go

```diff
@@ -77,28 +80,176 @@ func (k Keeper) ClaimCapability(ctx sdk.Context, cap *capabilitytypes.Capability
+	codeInfo, prefixStore, err := k.contractInstance(ctx, contractAddr)
+	if err != nil {
+		return err
+	}
+
+	var sender sdk.AccAddress // we don't know the sender
+	params := cosmwasm.NewEnv(ctx, sender, nil, contractAddr)
+	params.IBC = &ibcInfo
+
+	querier := QueryHandler{
+		Ctx:     ctx,
+		Plugins: k.queryPlugins,
+	}
+
+	gas := gasForContract(ctx)
+	mock, ok := MockContracts[contractAddr.String()] // hack for testing without wasmer
+	if !ok {
+		panic("not supported")
+	}
+	res, gasUsed, execErr := mock.OnAcknowledgement(codeInfo.CodeHash, params, payloadData, acknowledgement, prefixStore, cosmwasmAPI, querier, ctx.GasMeter(), gas)
+	consumeGas(ctx, gasUsed)
+	if execErr != nil {
+		return sdkerrors.Wrap(types.ErrExecuteFailed, execErr.Error())
+	}
+
+	// emit all events from this contract itself
+	events := types.ParseEvents(res.Log, contractAddr)
+	ctx.EventManager().EmitEvents(events)
+
+	// hack: use sdk messages here for simplicity
+	for _, m := range res.Messages {
+		if err := k.messenger.handleSdkMessage(ctx, contractAddr, m); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (k Keeper) OnTimeoutPacket(ctx sdk.Context, contractAddr sdk.AccAddress, payloadData []byte, ibcInfo cosmwasm.IBCInfo) error {
 	codeInfo, prefixStore, err := k.contractInstance(ctx, contractAddr)
 	if err != nil {
 		return err
 	}
-	params := types.NewEnv(ctx, data.Sender, nil, contractAddr)
 
-	// todo send data to contract
-	_, _, _ = codeInfo, prefixStore, params
+	var sender sdk.AccAddress // we don't know the sender
+	params := cosmwasm.NewEnv(ctx, sender, nil, contractAddr)
+	params.IBC = &ibcInfo
+
+	querier := QueryHandler{
+		Ctx:     ctx,
+		Plugins: k.queryPlugins,
+	}
+
+	gas := gasForContract(ctx)
+	mock, ok := MockContracts[contractAddr.String()]
+	if !ok { // hack for testing without wasmer
+		panic("not supported")
+	}
+	res, gasUsed, execErr := mock.OnTimeout(codeInfo.CodeHash, params, payloadData, prefixStore, cosmwasmAPI, querier, ctx.GasMeter(), gas)
+	consumeGas(ctx, gasUsed)
+	if execErr != nil {
+		return sdkerrors.Wrap(types.ErrExecuteFailed, execErr.Error())
+	}
+
+	// emit all events from this contract itself
+	events := types.ParseEvents(res.Log, contractAddr)
+	ctx.EventManager().EmitEvents(events)
+
+	// hack: use sdk messages here for simplicity
+	for _, m := range res.Messages {
+		if err := k.messenger.handleSdkMessage(ctx, contractAddr, m); err != nil {
+			return err
+		}
+	}
 	return nil
 }
 
-func (k Keeper) IBCCallContract(
-	ctx sdk.Context,
-	sourcePort,
-	sourceChannel string,
-	sender sdk.AccAddress,
-	receiver sdk.AccAddress,
-	timeoutHeight,
-	timeoutTimestamp uint64,
-	msg json.RawMessage,
-) error {
+func (k Keeper) IBCCallFromContract(ctx sdk.Context, sourcePort, sourceChannel string, sender sdk.AccAddress, timeoutHeight, timeoutTimestamp uint64, msg json.RawMessage) error {
```

Comment:
Cool idea, so this is another `CosmosMsg` type that the contract can return.

`IBCSendPacket` and `IBCOpenChannel` ?

Good to look at this code, I need to combine this with the comments on #214 and get a clear API spec'd out.

Reply:
I assume that Send and CloseChannel are possible from a contract. The `OpenChannel` process is not that clear to me! 

A channel is based on a connection and that on a client. Both are out of scope of our module so far. In GoS the relayer would establish them based on configuration. 
Technically the contract could store and provide the necessary data but calling out to other chains would mix with the relayers job and feels wrong at this stage.

from the cli:
` wasmcli tx ibc channel open-init [port-id] [channel-id] [counterparty-port-id] [counterparty-channel-id] [connection-hops] [flags]`

From the spec:
`The connectionHops stores the list of connection identifiers, in order, along which packets sent on this channel will travel. At the moment this list must be of length 1. In the future multi-hop channels may be supported.`